### DISCUSSION TITLE
Emitting emissions outside of actions/hooks

### DISCUSSION THREAD

**Discussion by zoton2** [RobinTail/zod-sockets#188] [Q&A] [answered]
Hey, been using express-zod-api for a bit on a couple of projects with good success, so thanks for that. I saw this project was being built too and was using Socket.IO too so wanted to try and use this library too.

Everything seems pretty good although I have one question: is it possible to send events out (emissions) to clients outside of places that are supplied with the relevant objects (like `all`), for example in actions or hooks? An extension of this is to also find clients from outside of these too.

Use case: an HTTP API request that is sent to the server that then looks up a client based on metadata and if found, sends an emission, waits for an acknowledgement, and returns said data to the HTTP request.

The way I did this before was to just store the clients in an external object alongside their metadata, but seeing as this library allows you to attach metadata, I wanted to give that a shot. This could be done again if it's the best solution (although typings might get a bit tricky).

The other way of doing this (that actually might be considered better, but I'm not sure?) would be to convert the HTTP API call to a (socket) action, where it would then have context. That or doing something with hooks, for example storing a copy of the parameter(s) from the `onStartup` hook.

I know it (emitting) can be done with the raw Socket.IO class but it has no typings. There may be an undocumented way of doing this too, so I thought I'd ask in case, or maybe I missed something in the documentation.

**Comment by RobinTail** [comment, marked as answer]
Hello @zoton2 ,

> an HTTP API request that is sent to the server that then looks up a client based on metadata

It's not supposed to work that way. HTTP server is only used as transport and for handshake to upgrade the connection to websockets. Moreover, "look up a client" means that there is a client, so that it is expected to be already connected via sockets. Therefore:

> The other way of doing this ... would be to convert the HTTP API call to a socket action

— that's exactly what I'd recommend to do.

So, you don't need to do any lookups. Your client emits the event (action), and you handle that action by emitting another event towards the original event owner (client). You'd not even need hooks for that. Very simple.
Thus, HTTP connection is only used to establish the socket connection once. Then, using that persistent bus you can send/receive as many events as you like.

Thank you for sharing your experience!

**Reply by zoton2** [reply]
> It's not supposed to work that way. HTTP server is only used as transport and for handshake to upgrade the connection to websockets. Moreover, "look up a client" means that there is a client, so that it is expected to be already connected via sockets.

Maybe something was misunderstood here. It's hard to explain as I can't show the project but the HTTP is independent of the sockets (using express-zod-api). There is an admin panel and also another part of the project which connects as the "clients". One example: originally I had a HTTP call that was then used in an admin panel that could respond with if a (socket) client was active or not, so it could show a "connected: true/false" for debug/admin purpose. Obviously this can either be done by just replacing the admin call with a socket connection, or storing that information in a (temporary) variable, [similar to what you do in your chat project](https://github.com/RobinTail/chat/blob/master/core/onTyping.ts), but the admin panel is currently fully HTTP so I wanted to check I wasn't missing something with the sockets.

There was some other lookups I was doing over HTTP to bridge the socket, like requesting information from the clients back over HTTP (to ask what they are currently doing) but that may also just be bad design in my project anyway.

I've also got the mindset that "big database call" *must* be over HTTP, which is probably a bad mindset to save.

Thanks for the quick reply!

**Reply by RobinTail** [reply]
I see. Okay, then you can use `all.getClients()` async method. It will return an array of all familiar clients. Then you can find one by `id` (which can be unstable) or using your own metadata (`.getData()` method).
Check out this article on how to define constraints and type assistance on metadata:
https://github.com/RobinTail/zod-sockets?tab=readme-ov-file#defining-constraints

@zoton2

**Reply by RobinTail** [reply]
But in order to use `all.getClients()` you should be within a hook. Perhaps you can emit some internal event for that, but I didn't explore that layer yet.

**Reply by RobinTail** [reply]
Ok, so far I can only suggest having some sort of shared `const` holding the client references to emit events to from a `onStartup` hook, @zoton2 
Maybe I'll invent a better built-in mechanism for that purpose.

**Reply by RobinTail** [reply]
Related topic: https://github.com/socketio/socket.io/issues/4724

**Reply by RobinTail** [reply]
🚀 v7.0.0 is published and offers the way to emit events anywhere in your application easily by returning an object of independent contexts from `attachSockets()`: https://github.com/RobinTail/zod-sockets/blob/main/README.md#elsewhere-in-your-application
@zoton2
